submission/proxy_server.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package submission
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/trillian/ctfe"
+)
+
+// ProxyServer wraps Proxy and handles http-requests for it.
+type ProxyServer struct {
+	p          *Proxy
+	addTimeout time.Duration
+}
+
+// NewProxyServer creates and inits ProxyServer instance.
+func NewProxyServer(logListPath string, logListRefreshInterval time.Duration, rootsRefreshInterval time.Duration, dBuilder DistributorBuilder, reqTimeout time.Duration) *ProxyServer {
+	s := &ProxyServer{addTimeout: reqTimeout}
+	s.p = NewProxy(NewLogListRefresher(logListPath), dBuilder)
+	s.p.Run(context.Background(), logListRefreshInterval, rootsRefreshInterval)
+	return s
+}
+
+func marshalSCTs(scts []*AssignedSCT) []byte {
+	var jsonSCTsObj struct {
```

Comment:
It would make much sense to extract this type into a named exported type, as it effectively is part of your public API, and clients will need to unmarshal it back.

Reply:
the type is not agreed upon yet. This one is more stub-type

Reply:
Not sure what stub-type means in this context. What I see is that this type is returned to clients via an HTTP handler. IMO it's a matter of good style to make this type non-ephemeral in the code. The fact that the API is not documented/final doesn't mean the type should be hidden in the internals of the code.

At the very minimum you could move the type declaration out of the function. At max something analogous to [types.go](https://github.com/google/certificate-transparency-go/blob/334df25da99ff4b12181480180d630e69ceb472f/types.go#L457) would be great.

Reply:
Done